### ISSUE TITLE
Issue with password sharing

### ISSUE THREAD

**Issue by Luticus** [nextcloud/passman#288]
### Steps to reproduce
Issue with password sharing to specific user 

(unsure, hoping for advice)

Oddly enough the user I cannot share with CAN share with me and I can share with other users.  We are both LDAP users and we have both regenerated our share keys.

When I click share the mesage says shared but it does not show the upload bars the way it does when it works with other users.  Instead it states that the share happened via a small popup at the top of the browser but no share notification is ever sent.  Shares that the other users sends me can be updated, modified, written to successfully.  Very odd situation.  Any ideas are greatly appreciated.  

### Configuration
**Operating system**:   
Debian 8 (both server and client)

**Browser**:      
chrome and firefox

**Passman version**:    
2.1.1

**Web server:**
apache 2

**Database:**
mysql 5.5

**PHP version:**
5      

**cloud server:** Nextcloud or ownCloud
nextcloud

**cloud version:** (see admin page or version.php)
11.0.2  
   
#### Browser log
<details>
<summary>Browser log</summary>

Insert your browser log here, this could for example include:

a) 
`passman.min.js?v=9360b83…:4 Error: Invalid PEM formatted message.
    at Object.t.decode (passman.min.js?v=9360b83…:18)
    at Object.n.publicKeyFromPem (passman.min.js?v=9360b83…:21)
    at passman.min.js?v=9360b83…:28
    at passman.min.js?v=9360b83…:4
    at m.$eval (passman.min.js?v=9360b83…:5)
    at m.$digest (passman.min.js?v=9360b83…:5)
    at m.$apply (passman.min.js?v=9360b83…:5)
    at l (passman.min.js?v=9360b83…:4)
    at J (passman.min.js?v=9360b83…:4)
    at XMLHttpRequest.t.onload (passman.min.js?v=9360b83…:4)
invalid {message: "json decode: this isn't json!", toString: function}`

</details>   

#### Cloud log
<details>
<summary>Cloud log</summary>
`{"reqId":"WOxHsH8AAQEAADpBuTIAAAAA","remoteAddr":"192.168.50.31","app":"PHP","message":"Only variables should be passed by reference at \/media\/cloud_data\/apps\/passman\/controller\/sharecontroller.php#235","level":3,"time":"2017-04-10T23:04:16-04:00","method":"DELETE","url":"\/index.php\/apps\/passman\/api\/v2\/sharing\/credential\/4E708547-C5B2-4582-BC11-3B79136A008E\/user2","user":"user1","version":"11.0.2.7"}
{"reqId":"WOxIG38AAQEAAD@R@zcAAAAI","remoteAddr":"192.168.50.31","app":"PHP","message":"vsprintf(): Too few arguments at \/var\/www\/owncloud\/lib\/private\/legacy\/l10n\/string.php#72","level":3,"time":"2017-04-10T23:06:05-04:00","method":"GET","url":"\/index.php\/apps\/passman\/api\/v2\/language?lang=en","user":"user1","version":"11.0.2.7"}
{"reqId":"WOxIG38AAQEAAD@R@zcAAAAI","remoteAddr":"192.168.50.31","app":"PHP","message":"Method OC_L10N_String::__toString() must return a string value at \/var\/www\/owncloud\/lib\/private\/L10N\/L10N.php#85","level":3,"time":"2017-04-10T23:06:05-04:00","method":"GET","url":"\/index.php\/apps\/passman\/api\/v2\/language?lang=en","user":"user1","version":"11.0.2.7"}
{"reqId":"WOxJ8H8AAQEAADYEatMAAAAL","remoteAddr":"192.168.50.31","app":"PHP","message":"vsprintf(): Too few arguments at \/var\/www\/owncloud\/lib\/private\/legacy\/l10n\/string.php#72","level":3,"time":"2017-04-10T23:13:52-04:00","method":"GET","url":"\/index.php\/apps\/passman\/api\/v2\/language?lang=en","user":"user1","version":"11.0.2.7"}
{"reqId":"WOxJ8H8AAQEAADYEatMAAAAL","remoteAddr":"192.168.50.31","app":"PHP","message":"Method OC_L10N_String::__toString() must return a string value at \/var\/www\/owncloud\/lib\/private\/L10N\/L10N.php#85","level":3,"time":"2017-04-10T23:13:52-04:00","method":"GET","url":"\/index.php\/apps\/passman\/api\/v2\/language?lang=en","user":"user1","version":"11.0.2.7"}
{"reqId":"WOxL238AAQEAAD1sZrsAAAAK","remoteAddr":"192.168.50.31","app":"PHP","message":"Only variables should be passed by reference at \/media\/cloud_data\/apps\/passman\/controller\/sharecontroller.php#235","level":3,"time":"2017-04-10T23:22:03-04:00","method":"DELETE","url":"\/index.php\/apps\/passman\/api\/v2\/sharing\/credential\/4E708547-C5B2-4582-BC11-3B79136A008E\/user2","user":"user1","version":"11.0.2.7"}`

</details>

**Comment by brantje** [comment]
I'm sorry for the late response. Kinda busy at work.
So, only ldap users are having trouble with sharing? #254 might be related in that case.
Will try to look into it this weekend, can't promise any thing because Easter.

**Comment by Luticus** [comment]
No prob, had a busy weekend myself, I'm happy you're looking into it at all.  

Not sure if it's just LDAP users or what because it appears to work for other LDAP users.  Was mostly hoping the errors might mean something to a dev and you guys could give me a direction to look in.  

If there's anything else you can think of that would help narrow it down let me know, I'd be happy to do whatever is needed to help figure out what's wrong.

Thanks!

**Comment by brantje** [comment]
From the browser log i can tell that the PEM cert is invalid. 
@animalillo is the share master here.

**Comment by Luticus** [comment]
Humm, odd.  You mean the PEM cert for apache?  That should be handled by let's encrypt if that's the case.  very strange.

I do have to set up a script to watch the hash's of the certs and reload apache when i changes every 3 months but that shouldn't have been needed yet.

**Comment by animalillo** [comment]
Hello @Luticus 
Sorry for the response times on this, been pretty busy lately.
If you want to contact me trough telegram i'm @WolFi there, feel free to pm me any time.

This is a weird issue and i might need you to do some tests for me on your setup.

Does the target user that causes trouble have more than 1 vault? Has him regenerated all the vault sharing keys from all the vaults? 

This looks a lot like one of the vaults have a corrupted certificate or no certificate at all on one of the vaults, as sharing sends the share request to every vault.

We should probably add a check and if sharing to one of the vaults fails warn about it with some popup or something, what do you think @brantje?

Also this issue has nothing to do with the apache certificate, or your browser would be complaining and the page won't load at all ^^

**Comment by Luticus** [comment]
Hi @animalillo ,

No problem.  Busy times here too  :)

As far as the vault situation we both have just one vault each.  originally the user had more vaults because they accidentally created a bunch.  So after the vault deletion process was put in place I had the user request all their vaults be removed and created one good one.  That does make me wonder if perhaps the vault deletion code is leaving behind some remnant in the database somewhere that's throwing the system?  Maybe something to look at.

Also not sure if regular DB users are treated differently then LDAP users some how, so I had thought that aspect might be worth looking into as well.  

As far as the certs, I'd thought I had regenerated the keys in the vaults for both users, but I could try again just to be sure.

For your warning, if you're going to implement that it would be great if the warning was sent to both parties and on the party with the bad vault key it stated which vault the delivery failed for. 

Yea, kinda figured the Apache cert wouldn't affect it.  shoulda guessed it was in regards to the vault keys  :)

I might see if I can get on telegram and look you up, don't have an account there yet.  

Thanks for looking into the issues and definitely let me know if there's any other info you need me to provide.

**Comment by Luticus** [comment]
Just thought I'd let you know that I updated to NC 11.0.3 today and tested this situation out again to see if by some chance it would help, but the issue remains.

**Comment by Luticus** [comment]
Just an FYI.

I solved this, it was as I suspected.  Finally had enough time set aside to check the database and in the passman_vaults table there was some crud entries left over from defunct vaults.  That was definitely affected sharing.  as soon as I cleared the bad records everything worked.  Might need to verify that the vaults are being cleaned up properly after removal or if this was just an odd fluke on my part.  I'll open a new issue if i discover the vaults aren't being cleaned up correctly after removal.
